src/Avalonia.Base/ValueStore.cs

```diff
@@ -178,5 +195,110 @@ public DeferredSetter<object> Setter
+                    {
+                        Array.Copy(_entries, 0, entries, 0, i);
+                    }
+
+                    entries[i] = new Entry { PropertyId = property.Id, Value = value };
+                    Array.Copy(_entries, i, entries, i + 1, _entries.Length - i);
+                    break;
+                }
+            }
+
+            _entries = entries;
+        }
+
+        private void SetValueInternal(AvaloniaProperty property, object value)
+        {
+            _entries[TryFindEntry(property.Id).Item1].Value = value;
+        }
+
+        private (int, bool) TryFindEntry(int propertyId)
+        {
+            if (_entries.Length <= 16)
+            {
+                // For small lists, we use an optimized linear search. Since the last item in the list
+                // is always int.MaxValue, we can skip a conditional branch in each iteration.
+                // By unrolling the loop, we can skip another unconditional branch in each iteration.
+
+                if (_entries[0].PropertyId >= propertyId) return (0, _entries[0].PropertyId == propertyId);
+                if (_entries[1].PropertyId >= propertyId) return (1, _entries[1].PropertyId == propertyId);
+                if (_entries[2].PropertyId >= propertyId) return (2, _entries[2].PropertyId == propertyId);
+                if (_entries[3].PropertyId >= propertyId) return (3, _entries[3].PropertyId == propertyId);
+                if (_entries[4].PropertyId >= propertyId) return (4, _entries[4].PropertyId == propertyId);
+                if (_entries[5].PropertyId >= propertyId) return (5, _entries[5].PropertyId == propertyId);
+                if (_entries[6].PropertyId >= propertyId) return (6, _entries[6].PropertyId == propertyId);
+                if (_entries[7].PropertyId >= propertyId) return (7, _entries[7].PropertyId == propertyId);
+                if (_entries[8].PropertyId >= propertyId) return (8, _entries[8].PropertyId == propertyId);
+                if (_entries[9].PropertyId >= propertyId) return (9, _entries[9].PropertyId == propertyId);
+                if (_entries[10].PropertyId >= propertyId) return (10, _entries[10].PropertyId == propertyId);
+                if (_entries[11].PropertyId >= propertyId) return (11, _entries[11].PropertyId == propertyId);
+                if (_entries[12].PropertyId >= propertyId) return (12, _entries[12].PropertyId == propertyId);
+                if (_entries[13].PropertyId >= propertyId) return (13, _entries[13].PropertyId == propertyId);
+                if (_entries[14].PropertyId >= propertyId) return (14, _entries[14].PropertyId == propertyId);
+            }
+            else
+            {
+                int low = 0;
+                int high = _entries.Length;
+                int id;
+
+                if (high > 0)
+                {
+                    while (high - low > 3)
+                    {
+                        int pivot = (high + low) / 2;
```

Comment:
You can divide by 2 using (x >> 1) operation. Seems it faster. This is macro optimisation, but may be it will be interesting.

Reply:
Any JIT will optimize that. We don't need to do that in our code.

Reply:
Oh, sorry. I didn't know and didn't check this

Reply:
No worries!